### PR TITLE
Add file encryption and decryption

### LINKED ISSUE #3: Create functions for encryption and decryption

### PR DESCRIPTION
**Pull request by raouft**
Basic file encryption/decryption using caesar cipher. Awaiting review.

### CODE REVIEW

**Review by brianzhouzc — DISMISSED** [review]
Missing decrypt function!

File: src/encrypt-decrypt.cpp [hunk 1]
```diff
@@ -0,0 +1,60 @@
+#include "encrypt-decrypt.h"
+
+using namespace std;
+
+char curr;
+
+// Random key to use for encryption/decryption
+int key = 214748;
+
+void encrypt(fs::path &input, fs::path &output) {
```
Comment (cppcooper) [inline src/encrypt-decrypt.cpp]: The functions should in theory encrypt or decrypt the file you pass it. I don't see the need for two paths. Unless of course we can come up with a use case we can implement which needs two.

Comment (raouft) [reply src/encrypt-decrypt.cpp]: That was my implementation initially but I added the two paths for increased flexibility and to make the process "non-destructive". I can change it back to in-place encryption/decryption if necessary though.

File: include/encrypt-decrypt.h [hunk 2]
```diff
@@ -0,0 +1,7 @@
+#pragma once
+#include "common.h"
+#include <iostream>
+
+void encrypt(fs::path &input, fs::path &output);
```
Comment (cppcooper) [inline include/encrypt-decrypt.h]: Functions defined in headers should be declared as extern. You can read about the keyword here <https://en.cppreference.com/w/cpp/language/storage_duration> or here <https://docs.microsoft.com/en-us/cpp/cpp/extern-cpp?view=msvc-160>. It is effectively used to avoid duplicate symbols.

File: src/encrypt-decrypt.cpp (hunk 1 above)
Comment (cppcooper) [reply src/encrypt-decrypt.cpp]: I would recommend it as it means we don't need to come up with secondary names for files.

File: include/encrypt-decrypt.h (hunk 2 above)
Comment (raouft) [reply include/encrypt-decrypt.h]: This has been added: e069834

File: src/encrypt-decrypt.cpp (hunk 1 above)
Comment (raouft) [reply src/encrypt-decrypt.cpp]: I've changed the functions to work in-place: 3e0d2eb

**Review by cppcooper — APPROVED** [review]
Everything looks okay.
edit: okay, on second look of the test this probably won't build currently

File: tests/encrypt-decrypt.cpp [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+    fs::path encryptedOutput("test-files/file_encrypted.txt");
+    fs::path decryptedOutput("test-files/file_decrypted.txt");
+
+    encrypt(input, encryptedOutput);
+    assert(!is_data_duplicate(input, encryptedOutput));
+
+    decrypt(encryptedOutput, decryptedOutput);
+    assert(is_data_duplicate(input,decryptedOutput));
+    std::cout << "test passed\n";
+}
```
Comment (cppcooper) [inline tests/encrypt-decrypt.cpp]: Clever test by the way, using the is_data_duplicate like this.
This does now block the PR though, as you'll need to come up with an update for the test.

Comment (raouft) [reply tests/encrypt-decrypt.cpp]: Thanks! 

And yeah, completely forgot to update the test - working on it now..

Comment (raouft) [reply tests/encrypt-decrypt.cpp]: Updated: d8757c9

File: tests/encrypt-decrypt.cpp [hunk 4]
```diff
@@ -9,10 +9,11 @@ int main(){
     fs::path encryptedOutput("test-files/file_encrypted.txt");
     fs::path decryptedOutput("test-files/file_decrypted.txt");
 
-    encrypt(input, encryptedOutput);
-    assert(!is_data_duplicate(input, encryptedOutput));
+    encrypt(input);
+    assert(is_data_duplicate(input, encryptedOutput));
```
Comment (cppcooper) [inline tests/encrypt-decrypt.cpp:13]: This still checks different files. I think you'll need to copy the original file to a second path and compare those two files to avoid changing the test too much.

Or you could always go back to your original function implementation, and we can feed the same file in twice later on, you'll need to test that edge case too though to make sure it correctly rewrites the file since you'll have two fstreams using the same file.

**Review by cppcooper — CHANGES_REQUESTED** [review]

File: tests/encrypt-decrypt.cpp (hunk 4 above)
Comment (raouft) [reply tests/encrypt-decrypt.cpp:13]: That's what I did - copied the original file into file_decrypted.txt. Or are you saying I should copy the file programmatically as part of the test?

I also put the expected output in file_encrypted.txt so I could compare it with the output after running encrypt().

Comment (cppcooper) [reply tests/encrypt-decrypt.cpp:13]: Yes I meant programmatically. But my thinking was you wouldn't be testing what you think you are.. But I suppose those files will be static. It'll just break the test if the original file is modified. You make a decision, resolve this convo when you've made up your mind and merge when ready. I think there is nothing left to gripe about after this.

Comment (raouft) [reply tests/encrypt-decrypt.cpp:13]: I see the issue you're referring to but I think it's fine to leave like this for the sake of this simple exercise

**Comment by cppcooper** [comment]
Well I learned a value lesson about change requests. I'll disable it, so we can merge and avoid this mistake in the future.
